File: data/java_template/blockchain/src/main/java/beans/Block.java
package beans;


import entities.Blockchain;
import entities.Transaction;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;

/**
 * Block class represents the basic part of the blockchain
 *
 * Implements the Serializable inteface in order to be sent above
 * network when a new miner joins the blockchain network
 */
public class Block implements Serializable {

                                       
    private String previousHash;
    private long timestamp;
    private String hash;
    private int nonce;
    private List<Transaction> transactions = new ArrayList<Transaction>();

    /*
     * todo:
     * Function that calculates the hash on the current block
     */
    public String calculateHash() throws Exception {

        return "";
    }

    /*
     * todo:
     * Function that adds a Transaction on the current block if it is valid
     */
    public boolean addTransaction(Transaction transaction, Blockchain blockchain) {

        return true;
    }

}

File: src/block.js
const hash = require('object-hash');

module.exports = class BLOCK {
    index;
    timestamp;
    transactions;
    nonce;
    current_hash;
    previous_hash;
    constructor(index, nonce, previous_hash) {
        this.index = index;
        this.timestamp = Date.now();
        this.transactions = [];
        this.nonce = nonce;
        this.current_hash = this.hashBlock();
        this.previous_hash = previous_hash;
    }

    hashBlock() {
        var blockdata = {
            index:          this.index,
            timestamp:      this.timestamp,
            transactions:   this.transactions,
            nonce:          this.nonce
        }
        return hash(blockdata);
    }
}